src/Uno.UI/UI/Xaml/ResourceDictionary.cs

```diff
@@ -48,7 +72,41 @@ public Uri Source
 		}
 
 		public IList<ResourceDictionary> MergedDictionaries => _mergedDictionaries;
-		public IDictionary<object, object> ThemeDictionaries => _themeDictionaries ??= new ResourceDictionary();
+		public IDictionary<object, object> ThemeDictionaries => GetOrCreateThemeDictionaries();
+
+		private ResourceDictionary GetOrCreateThemeDictionaries()
+		{
+			if (_themeDictionaries is null)
+			{
+				_themeDictionaries = new ResourceDictionary();
+				_themeDictionaries.ResourceDictionaryValueChange += (sender, e) =>
+				{
+					// The purpose of this to updated the cached value when it's no longer valid.
+					// See GetActiveThemeDictionary.
+					if (e.ChangedKey is ResourceKey changedKey && changedKey.Equals(_activeTheme))
+					{
+						if (e.NewValue is null && !_activeTheme.Equals(Themes.Default))
+						{
+							// If the key is removed and we have non-default active theme,
+							// then update cached value with default theme.
+							// We do that here because GetActiveThemeDictionary does that.
+							_activeThemeDictionary = GetThemeDictionary(Themes.Default);
```

Comment:
Since you're now setting _activeTheme from GetThemeDictionary(), this will set _activeTheme to Default, which seems strange?

Instead of trying to recalculate _activeThemeDictionary here, can you simply set _activeTheme to ResourceKey.Empty? Then _activeThemeDictionary will be correctly re-resolved the next time `GetActiveThemeDictionary()` is called.

Reply:
Doing so is still problematic for the following case:

- Active theme is Light but we only have Default
- We assign `Light` to `_activeTheme`, and the default dictionary to `_activeThemeDictionary` (through the `?? GetThemeDictionary(Themes.Default)`
- Later, "Default" gets removed

For this scenario, we shouldn't have anything available, but we'll be caching the "Default" theme if the `_activeTheme` wasn't set to `Default`. Is there a better approach that handles this scenario?

Reply:
But it's problematic in the case that we have cached the "Default" dictionary, and later on "Light" gets added, or no?

Can you just invalidate `_activeTheme` when _any_ dictionary is added/removed from ThemeDictionaries? That seems safest, and modifying `ThemeDictionaries` is sufficiently uncommon that it hopefully shouldn't pose a performance problem.

Reply:
> But it's problematic in the case that we have cached the "Default" dictionary, and later on "Light" gets added, or no?

In that case, the next time `GetActiveThemeDictionary` is called won't use the cached value because `_activeTheme` is "Default" but we're requesting "Light". So it shouldn't be problematic?

Anyway I agree that invalidating _activeTheme whenever a dictionary is added or removed is safest. But I want to make sure we have a good test coverage in case someone adds the caching back in future :)

Reply:
> In that case, the next time GetActiveThemeDictionary is called won't use the cached value because _activeTheme is "Default" but we're requesting "Light". So it shouldn't be problematic?

It sounds like it's the caching itself that would be broken in that case, but I can't see the code anymore, so never mind. :D

And yes, whatever the fix, good test coverage is useful. ;)

Reply:
@davidjohnoliver Do you mean the current implementation is okay, or I should still be invalidating the cache for any theme dictionary change? or both are okay (in which case I'd like to see if @jeromelaban has any preference)?